torch/dispatch/dispatcher.py

```diff
@@ -0,0 +1,95 @@
+import torch
+import torch._C as _C
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard  # type: ignore[attr-defined]
+from torch.utils._pytree import tree_flatten
+from torch.overrides import handle_torch_function, has_torch_function
+
+SUPPORTED_KEYS = {
+    DispatchKey.CPU,
+    DispatchKey.BackendSelect,
+    DispatchKey.ADInplaceOrView,
+    DispatchKey.AutogradCPU,
+    DispatchKey.Python,
+    DispatchKey.PythonTLSSnapshot,
+}
+
+"""
+This is a dispatcher (in Python)
+- You can define new operations (in Python) without schemas
+- It interfaces with the PyTorch dispatcher
+"""
+
+class PyDispatcher:
+    def __init__(self):
+        self.current_dispatching_op = None
+        self.already_dispatched_keys = None
+        self.current_key = None
+
+    def call(self, operator, args, kwargs):
+        try:
+            key = compute_dispatch_key(operator, args, kwargs, self.current_key)
+            self.record_dispatch(key, operator)
+            return dispatch(key, operator, args, kwargs)
+        finally:
+            self.reset_dispatch_record()
+
+    def redispatch(self, operator, args, kwargs):
+        # Redispatch doesn't go to the top
+        assert operator == self.currently_dispatching_op
+        key = compute_dispatch_key(operator, args, kwargs, self.current_key, self.already_dispatched_keys)
```

Comment:
This isn't faithful. redispatch in C++ takes a DispatchKeySet which specifies what to dispatch to next; it's not implicitly computed.

Reply:
I wrote redispatch and stored the state mentioned in another comment as a quick hack when I was trying to get this to work, but yeah, we should be faithful to how the dispatcher works

Reply:
Sure, I can make this change.